File: BinaryTrees/BalancedBT/balance.cpp
#include <iostream>
using namespace std;

class Node {
  public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = right = nullptr;
    }
};


class Solution {
  public:
    int findHeight(Node* root, int level){
        if(!root)return 0;
        return 1+max(findHeight(root->left,level+1),findHeight(root->right,level+1));
    }
    bool isBalanced(Node* root) {
        // code here
        if(!root)return true;
        int leftHeight = findHeight(root->left,0);
        int rightHeight = findHeight(root->right,0);
        int diffHeight = abs(leftHeight-rightHeight);
        if(diffHeight > 1)return false;
        return isBalanced(root->left) && isBalanced(root->right);
    }
};

File: BinaryTrees/BalancedBT/balance.py
class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None

class Solution:
    def findHeight(self, root, level):
        if root is None:
            return 0

        return 1 + max(self.findHeight(root.left, level + 1),
                       self.findHeight(root.right, level + 1))

    def isBalanced(self, root):
        if root is None:
            return True

        leftHeight = self.findHeight(root.left, 0)
        rightHeight = self.findHeight(root.right, 0)

        diffHeight = abs(leftHeight - rightHeight)

        if diffHeight > 1:
            return False

        return (self.isBalanced(root.left) and
                self.isBalanced(root.right))